Source/Core/Core/PowerPC/JitArm64/JitArm64_Paired.cpp

```diff
@@ -110,46 +110,136 @@ void JitArm64::ps_maddXX(UGeckoInstruction inst)
   ARM64Reg VB = reg_encoder(fpr.R(b, type));
   ARM64Reg VC = reg_encoder(fpr.R(c, type));
   ARM64Reg VD = reg_encoder(fpr.RW(d, type));
-  ARM64Reg V0Q = fpr.GetReg();
-  ARM64Reg V0 = reg_encoder(V0Q);
-
-  // TODO: Do FMUL and FADD/FSUB in *one* host call to save accuracy.
+  ARM64Reg V0Q = INVALID_REG;
+  ARM64Reg V0 = INVALID_REG;
+  if ((d != b || op5 == 28) && (d == a || d == c))
+  {
+    V0Q = fpr.GetReg();
+    V0 = reg_encoder(V0Q);
+  }
 
   switch (op5)
   {
   case 14:  // ps_madds0
-    m_float_emit.FMUL(size, V0, VA, VC, 0);
-    m_float_emit.FADD(size, VD, V0, VB);
+    // d = a * c.ps0 + b
+    if (d == b)
+    {
+      m_float_emit.FMLA(size, VD, VA, VC, 0);
+    }
+    else if (d != a && d != c)
+    {
+      m_float_emit.MOV(VD, VB);
+      m_float_emit.FMLA(size, VD, VA, VC, 0);
+    }
+    else
+    {
+      m_float_emit.MOV(V0, VB);
+      m_float_emit.FMLA(size, V0, VA, VC, 0);
+      m_float_emit.MOV(VD, V0);
+    }
     break;
   case 15:  // ps_madds1
-    m_float_emit.FMUL(size, V0, VA, VC, 1);
-    m_float_emit.FADD(size, VD, V0, VB);
+    // d = a * c.ps1 + b
+    if (d == b)
+    {
+      m_float_emit.FMLA(size, VD, VA, VC, 1);
+    }
+    else if (d != a && d != c)
+    {
+      m_float_emit.MOV(VD, VB);
+      m_float_emit.FMLA(size, VD, VA, VC, 1);
+    }
+    else
+    {
+      m_float_emit.MOV(V0, VB);
+      m_float_emit.FMLA(size, V0, VA, VC, 1);
+      m_float_emit.MOV(VD, V0);
+    }
     break;
   case 28:  // ps_msub
-    m_float_emit.FMUL(size, V0, VA, VC);
-    m_float_emit.FSUB(size, VD, V0, VB);
+    // d = a * c - b
+    if (d != a && d != c)
+    {
+      m_float_emit.FNEG(size, VD, VB);
+      m_float_emit.FMLA(size, VD, VA, VC);
+    }
+    else
+    {
+      m_float_emit.FNEG(size, V0, VB);
+      m_float_emit.FMLA(size, V0, VA, VC);
+      m_float_emit.MOV(VD, V0);
```

Comment:
Another thing to consider for the `d == b` case:```
      m_float_emit.FMLS(size, VD, VA, VC);
      m_float_emit.FNEG(size, VD, VD);
```
Note however this would not be accurate when the fp rounding mode is to +/- infinity.

Reply:
Thanks, added. I'm already not handling +/- infinity rounding modes correctly, so I don't mind adding another case.